Fix #6004: Added a presubmit check to ensure imports of JS files are sorted

anubhavsinha98 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #6004 : This PR consists of fixing the imports of the js files, and adding a presubmit check to ensure same in future.
 
## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

Commit:
First

Commit:
First

Commit:
Second

Commit:
Third

Commit:
Fourth

anubhavsinha98 committed:
Fixed imports order

anubhavsinha98 committed:
Few more

anubhavsinha98 committed:
1800

anubhavsinha98 committed:
1900

anubhavsinha98 committed:
Removed

anubhavsinha98 committed:
corrected

anubhavsinha98 committed:
2010

anubhavsinha98 committed:
Final

anubhavsinha98 committed:
resolved linting issues

anubhavsinha98 committed:
resolved linting issues

anubhavsinha98 commented:
@apb7 , @seanlip PTAL! Thanks!

anubhavsinha98 commented:
After merging this PR then we can move to this [PR](https://github.com/oppia/oppia/pull/6113)

anubhavsinha98 committed:
Fixed HintsAndSolution

DubeySandeep commented:
Hi @anubhavsinha98, FYI, this PR is failing Travis tests PTAL!

anubhavsinha98 commented:
Hi @DubeySandeep, I have fixed all the imports correctly (i.e. sorted them) . I checked Travis CI build log I found a common error which failed many tests i.e.

![screenshot from 2019-01-19 13-39-03](https://user-images.githubusercontent.com/30555544/51424135-bf2c5c80-1bef-11e9-8bd8-cf227e108521.png)

Can you please take a look and suggest something as I am not able to recognise why this error is prevailing?

DubeySandeep commented on core/templates/dev/head/pages/exploration_player/HintAndSolutionModalService.js:
```diff
@@ -17,17 +17,19 @@
+  'AutogeneratedAudioPlayerService',
+  'HintsAndSolutionManagerService',
+  'PlayerPositionService', 'PlayerTranscriptService',
+  'ServiceAudioPlayerService', 'UrlInterpolationService', 'COMPONENT_NAME_HINT',
```
@anubhavsinha98, `ServiceAudioPlayerService` do we need this? (also, do we have any service like this?)

How I found this:
- You'll find a link for each error in angularJS which will explain what's going wrong, I found this [link](https://code.angularjs.org/1.5.8/docs/error/$injector/unpr?p0=ServiceAudioPlayerServiceProvider%20%3C-%20ServiceAudioPlayerService%20%3C-%20HintAndSolutionModalService) over [here](https://travis-ci.org/oppia/oppia/jobs/480758947#L5418).

anubhavsinha98 replied:
Oh really thanks @DubeySandeep , actually it's my fault!

anubhavsinha98 committed:
Fixed Hintandsolution

anubhavsinha98 commented:
Hi @apb7 @DubeySandeep  all the tests have been passed PTAL !

anubhavsinha98 committed:
Merge branch 'develop' into test

apb7 commented on core/templates/dev/head/components/ExplorationCreationService.js:
```diff
@@ -18,11 +18,12 @@
-      AlertsService, SiteAnalyticsService, UrlInterpolationService) {
+      $http, $rootScope, $timeout, $uibModal, $window,
+      AlertsService, SiteAnalyticsService, UrlInterpolationService
+  ) {
```
Undo this change -- move the brackets to the previous line.

anubhavsinha98 replied:
Okay will do that !

apb7 reviewed (commented):
Hi @anubhavsinha98, this is looking almost good. Also, I was thinking if we could incorporate changes made in #6113 in this PR and close that one. In this way, we will be able to verify that all cases have been covered and see the check in action. Thanks!

anubhavsinha98 committed:
Fixed ExplorationCreationService.js

anubhavsinha98 committed:
Merge branch 'test' of https://github.com/anubhavsinha98/oppia into test

anubhavsinha98 committed:
Added presubmit check

anubhavsinha98 commented:
Hi @apb7 added the check also, PTAL!  Thanks!

anubhavsinha98 committed:
Fixed all files

anubhavsinha98 commented:
Hi @apb7, all tests have been passed, PTAL !

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1439,6 +1439,105 @@ def _check_directive_scope(all_files):
     return summary_messages
 
 
+def _check_sorted_imports(all_files):
```
The function is name is too general, please rename it to something concrete.

anubhavsinha98 replied:
Done.

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1439,6 +1439,105 @@ def _check_directive_scope(all_files):
 
+def _check_sorted_imports(all_files):
+    """This function checks that the imports are sorted
+    according to di-order rule of  angular plugin of eslint.
```
Nit: Extra space after `of`

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1439,6 +1439,105 @@ def _check_directive_scope(all_files):
 
 
+def _check_sorted_imports(all_files):
+    """This function checks that the imports are sorted
```
Also, vague description. And this isn't sorting based on di-order rule, its a custom sorting. (If that would have been the case, we wouldn't require this :))

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1439,6 +1439,105 @@ def _check_directive_scope(all_files):
+    """This function checks that the imports are sorted
+    according to di-order rule of  angular plugin of eslint.
+    """
+    print 'Starting sorted imports check'
```
Ditto, this would make difficult for the user to distinguish between python import sorting check and this one.

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1439,6 +1439,105 @@ def _check_directive_scope(all_files):
+                regular_imports.sort()
+                constant_imports.sort()
+                sorted_imports = (dollar_imports + regular_imports +
+                                  constant_imports) #pylint: disable=line-too-long
```
No pylint disables; break this line up

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1439,6 +1439,105 @@ def _check_directive_scope(all_files):
+    for filename in files_to_check:
+        with open(filename) as f:
+            content = f.read()
+        parsed_dict = _validate_and_parse_js_file(filename, content)
```
I think we are also parsing this in another directive check. How about we do this once and share the dict between the two functions? Maybe populate the dict only once and share it. Will save us some time for sure!

anubhavsinha98 replied:
Yeah, this will be good.

anubhavsinha98 replied:
How about if we make a separate function for validating and parsing JS files, then share its content(in the form of a dictionary)by passing it in the function?
What do you suggest?

anubhavsinha98 replied:
Done

apb7 reviewed (changes requested):
@anubhavsinha98, left some comments. This is getting close. Also, can you double check that no dependency has been missed while re-arranging it, before the next review? Thanks!

anubhavsinha98 committed:
Added one time parsing only

anubhavsinha98 committed:
Made some changes

anubhavsinha98 commented:
Hi @apb7 done with the changes PTAL ! Thanks.

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/anubhavsinha98/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' into test

anubhavsinha98 commented:
Hi @apb7, PTAL! Thanks.

DubeySandeep commented:
@apb7, can you please review this PR?

anubhavsinha98 committed:
Merge branch 'develop' into test

anubhavsinha98 committed:
Update pre_commit_linter.py

anubhavsinha98 commented:
Hi @apb7, removed the merge conflict. PTAL! Thanks.

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1322,17 +1322,27 @@ def _check_html_directive_name(all_files):
+        filename for filename in all_files if not
+        any(fnmatch.fnmatch(filename, pattern) for pattern in EXCLUDED_PATHS)
+        and filename.endswith('.js')]
+    for filename in files_to_check:
```
Use the existing file caching here (see [this line](https://github.com/oppia/oppia/pull/6125/files#diff-3093ff31fd2fc587c0d99fca7d3afdf0L1350))

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1472,106 @@ def _check_directive_scope(all_files):
 
+def _check_sorted_dependencies(all_files, parsed_js_files_dict):
+    """This function checks that the dependencies which are
+    imported in the controllers/directives/factories in js
```
`js` --> `JS`. Also, can you please check if more words can be accommodated in each line?

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1472,106 @@ def _check_directive_scope(all_files):
+def _check_sorted_dependencies(all_files, parsed_js_files_dict):
+    """This function checks that the dependencies which are
+    imported in the controllers/directives/factories in js
+    files are in following pattern, first dollar imports in
```
Re-phrase this like: `..following pattern: dollar imports, regular imports and constant imports, all in sorted form.`

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1472,106 @@ def _check_directive_scope(all_files):
+                    print (
+       	                'Please ensure that %s in %s, the dependecies literally'
+       	                ' mentioned should be in the following manner i.e '
+       	                'Firstly dollar imports in sorted form, then regular '
```
Rephrase these messages too (see earlier comment).

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1472,106 @@ def _check_directive_scope(all_files):
+    properties_to_check = ['controller', 'directive', 'factory']
+    failed = False
+    summary_messages = []
+    for filename in files_to_check:
```
Processing and printing should be done within target stdout: `with _redirect_stdout(_TARGET_STDOUT):`
See other checks for reference.

anubhavsinha98 replied:
Done

anubhavsinha98 replied:
Done

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1773,7 +1882,11 @@ def main():
     all_files = _get_all_files()
     linter_messages = _pre_commit_linter(all_files)
-    directive_scope_messages = _check_directive_scope(all_files)
+    parsed_js_files_dict = _validate_and_parse_js_file(all_files)
```
`_validate_and_parse_js_file` --> `_validate_and_parse_js_files`

anubhavsinha98 replied:
Done

anubhavsinha98 committed:
changes done-linitng issues fixed

anubhavsinha98 commented:
Hi @apb7, changes are done. PTAL! Thanks!

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1322,17 +1322,26 @@ def _check_html_directive_name(all_files):
-    """This function validates a JavaScript file and returns the parsed contents
-    as a Python dictionary.
+def _validate_and_parse_js_files(all_files):
+    """This function validates a JavaScript files and returns the parsed
```
~a~

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1471,106 @@ def _check_directive_scope(all_files):
+                    if sorted_imports != function_args:
+                        failed = True
+                        print (
+                            'Please ensure that %s in %s, the injected '
```
`Please ensure that in %s in file %s, ....`

anubhavsinha98 replied:
Done

apb7 replied:
Not done? Please read the comment carefully before "Done"ing them.

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1471,106 @@ def _check_directive_scope(all_files):
+                        failed = True
+                        print (
+                            'Please ensure that %s in %s, the injected '
+                            'dependecies should be in the following manner:'
```
Spell check, dependencies.

anubhavsinha98 replied:
Done

apb7 replied:
Not done??

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1471,106 @@ def _check_directive_scope(all_files):
+                            'Please ensure that %s in %s, the injected '
+                            'dependecies should be in the following manner:'
+                            ' dollar imports, regular imports and constant '
+                            'imports, all in sorted form.'
```
`form` --> `order`

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1471,106 @@ def _check_directive_scope(all_files):
+                    if sorted_imports != literal_args:
+                        failed = True
+                        print (
+       	                    'Please ensure that %s in %s, the dependecies '
```
Similar comments here. Also, what's `literally` here? It seems to be out of context.

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1471,106 @@ def _check_directive_scope(all_files):
+        summary_message = (
+            '%s   Sorted dependencies check failed' % (
+                _MESSAGE_TYPE_FAILED))
+        print summary_message
```
Pull this and the next line out: they are being repeated in both the blocks.

anubhavsinha98 replied:
Done

anubhavsinha98 committed:
Fixed the changes

anubhavsinha98 commented:
Hi @apb7, PTAL! Thanks.

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1540,30 +1540,30 @@ def _check_sorted_dependencies(all_files, parsed_js_files):
+            summary_message = (
+                '%s  Sorted dependencies check failed' % (
+                    _MESSAGE_TYPE_FAILED))
+            summary_messages.append(summary_message)
```
Read the previous comment _carefully_: pull out this line as well.

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1540,30 +1540,30 @@ def _check_sorted_dependencies(all_files, parsed_js_files):
-       	                    'literally mentioned should be in the following '
-       	                    ' manner: dollar imports, regular imports and '
-       	                    'constant imports, all in sorted form.'
+       	                    'Please ensure that %s in %s, the stringfied '
```
Read the previous comment _carefully_!

anubhavsinha98 replied:
Hi @apb7 is the error message correct, or needs to be modified ?

apb7 reviewed (changes requested):
Hi @anubhavsinha98, please do not be in a hurry when addressing review comments. Also, before asking for another pass, ensure that _all_ comments are addressed _completely_. Thanks!

anubhavsinha98 commented:
> Hi @anubhavsinha98, please do not be in a hurry when addressing review comments. Also, before asking for another pass, ensure that _all_ comments are addressed _completely_. Thanks!

Hi @apb7, sorry for this, will not repeat again.

anubhavsinha98 committed:
Updated the changes

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1537,15 +1537,15 @@ def _check_sorted_dependencies(all_files, parsed_js_files):
                         print (
-                            'Please ensure that %s in %s, the injected '
-                            'dependecies should be in the following manner:'
+                            'Please ensure that in %s in %s, the injected '
```
Okay, I am re-iterating the earlier comment here: `Please ensure that in %s in file %s, ....`. Please change this here and below.

anubhavsinha98 replied:
Done.

anubhavsinha98 committed:
Merge branch 'develop' into test

anubhavsinha98 committed:
Changes done

anubhavsinha98 commented:
Hi @apb7, changes done. PTAL!

apb7 commented:
@anubhavsinha98, failing checks! Please do not use GitHub's online editor for making changes -- it bypasses the pre-submit checks. Thanks!

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

apb7 commented:
@anubhavsinha98, updates on this? Thanks!

anubhavsinha98 commented:
Hi @apb7, will update you soon, after removing the merge conflicts.

anubhavsinha98 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

anubhavsinha98 committed:
Resolved Conflicts

anubhavsinha98 committed:
Fixed linting

brianrodri reviewed (approved):
This is amazing!!! Thanks so much 😄

anubhavsinha98 committed:
Fixed failed travis ci check

anubhavsinha98 commented:
Hi @apb7, done with changes and removed the conflicts. PTAL. Thanks!

anubhavsinha98 commented:
> This is amazing!!! Thanks so much smile

Welcome @brianrodri! :) :+1:

brianrodri commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1471,107 @@ def _check_directive_scope(all_files):
+def _check_sorted_dependencies(all_files, parsed_js_files):
+    """This function checks that the dependencies which are
+    imported in the controllers/directives/factories in JS
+    files are in following pattern: dollar imports ,regular
```
Bad comma placement, please fix:

```
files are in following pattern: dollar imports ,regular
```

Should be:

```
files are in following pattern: dollar imports, regular
```

anubhavsinha98 replied:
Thanks for picking out this!
Done.

brianrodri commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1471,107 @@ def _check_directive_scope(all_files):
+    """This function checks that the dependencies which are
+    imported in the controllers/directives/factories in JS
+    files are in following pattern: dollar imports ,regular
+    imports and constant imports, all in sorted order.
```
Use oxford comma:

```
imports and constant imports, all in sorted order.
```

Should be:

```
imports, and constant imports, all in sorted order.
```

anubhavsinha98 replied:
Done.

brianrodri commented on scripts/pre_commit_linter.py:
```diff
@@ -1463,6 +1471,107 @@ def _check_directive_scope(all_files):
+    files_to_check = [
+        filename for filename in all_files if not
+        any(fnmatch.fnmatch(filename, pattern) for pattern in EXCLUDED_PATHS)
+        and filename.endswith('.js')]
```
This is a bit hard to read, please make `filename.endswith('.js')` the first condition checked to improve!

anubhavsinha98 replied:
Done

brianrodri commented on scripts/pre_commit_linter.py:
```diff
@@ -1322,17 +1322,26 @@ def _check_html_directive_name(all_files):
+    files_to_check = [
+        filename for filename in all_files if not
+        any(fnmatch.fnmatch(filename, pattern) for pattern in EXCLUDED_PATHS)
+        and filename.endswith('.js')]
```
Same as other comment, please fix!

anubhavsinha98 replied:
Done

anubhavsinha98 committed:
Fixed changes

anubhavsinha98 committed:
Fixed changes

anubhavsinha98 committed:
Fixed linting issues

kevinlee12 commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -27,19 +27,23 @@ oppia.constant('LIBRARY_PAGE_MODES', {
-  'LIBRARY_PAGE_MODES', 'LIBRARY_TILE_WIDTH_PX', 'AlertsService',
-  'LearnerDashboardIdsBackendApiService', 'UserService',
-  'LearnerDashboardActivityIdsObjectFactory', 'LearnerPlaylistService',
+  '$http', '$log', '$rootScope', '$scope', '$timeout', '$uibModal', '$window',
```
If I'm not mistaken, I believe `$scope` should come first, then the rest comes in alphabetical order.

anubhavsinha98 replied:
Hi @kevinlee12, as @brianrodri suggested [here](https://github.com/oppia/oppia/issues/6004#issue-392503295). All the sorted dollar imports should come first, then regular ones, and then the constant ones.

kevinlee12 replied:
noted, though I'm pointing this out since it's a (implicit) convention that is typically followed, so this would change that convention.

seanlip replied:
/cc @oppia/dev-workflow-team 

What should the convention be? We should finalize the decision on it, and have a lint check to enforce it.

Also, note that this shouldn't block merge of this PR. We should fix all the occurrences once we introduce the convention.

anubhavsinha98 replied:
Hi @seanlip, I have refactor all the files to introduce this convention.

seanlip replied:
Just to be clear, I am fine with any convention that the dev workflow team agrees on. I just want to suggest that the convention be clearly stated [here](https://github.com/oppia/oppia/wiki/Coding-style-guide), and that there's a lint check for it.

@apb7, as lead of the dev workflow team, could you please weigh in?

apb7 replied:
@seanlip, I am fine with this convention as stated here: https://github.com/oppia/oppia/issues/6004#issue-392503295. This PR refactors the files as well as adds a lint check for the above-mentioned convention.
@anubhavsinha98, can you please add this to the coding style guide here: https://github.com/oppia/oppia/wiki/Coding-style-guide#javascript
Thanks!

anubhavsinha98 replied:
Okay sure @apb7 !

anubhavsinha98 replied:
Done

apb7 commented on scripts/pre_commit_linter.py:
```diff
@@ -1495,11 +1604,10 @@ def _match_line_breaks_in_controller_dependencies(all_files):
-                        'and the dependencies mentioned as function parameters:'
-                        '\n(%s)\nfor the corresponding controller should '
-                        'exactly match.' % (
+                        '%s --> Please ensure that line breaks between '
```
Why this change? Please take care while merging your local branch with develop.

anubhavsinha98 replied:
Sorry for this. Corrected it.

apb7 reviewed (commented):
@anubhavsinha98, ptal!

anubhavsinha98 committed:
Changes done

anubhavsinha98 commented:
Hi @apb7, added the convention in the coding style guide.
![screenshot from 2019-02-20 19-59-21](https://user-images.githubusercontent.com/30555544/53098438-9dedc180-3549-11e9-8741-7a3a72766010.png)

anubhavsinha98 commented:
Hi @apb7, changes are done! PTAL. Thanks!

apb7 reviewed (approved):
LGTM, thanks @anubhavsinha98 :)

bansalnitish reviewed (approved):
LGTM, Thanks @anubhavsinha98!

apb7 commented:
@aks681, @nithusha21, @DubeySandeep, @kevinlee12, @seanlip: This requires your code owner's approval. Thanks!

nithusha21 reviewed (approved):
LGTM for changes in files relating to suggestions/feedback.

seanlip reviewed (approved):
LGTM

kevinlee12 reviewed (approved):
LGTM!

aks681 reviewed (approved):
LGTM!

anubhavsinha98 committed:
Merge branch 'develop' into test

anubhavsinha98 commented:
Hi @DubeySandeep, PTAL. Thanks!

anubhavsinha98 committed:
Updated LocalNav.js

anubhavsinha98 commented:
Hi @apb7, Travis CI checks have been passed.

apb7 commented:
Great work @anubhavsinha98 :)
